### PR TITLE
Allow customization of UUID encoding and columns

### PR DESCRIPTION
**Pull request by shadowhand**
Fixes https://github.com/EventSaucePHP/EventSauce/issues/138

### CODE REVIEW

File: src/DoctrineMessageRepository/DoctrineUuidV4MessageRepository.php [hunk 1]
```diff
@@ -19,99 +19,35 @@
...
+        MessageSerializer $serializer,
+    ) {
+        $this->messageRepository = new DoctrineMessageRepository(
+            connection: $connection,
+            serializer: $serializer,
+            tableSchema: new DefaultTableSchema(
+                tableName: $tableName
+            ),
+            uuidSerializer: new BinaryUuidSerializer(),
+        );
```
Comment (shadowhand) [inline src/DoctrineMessageRepository/DoctrineUuidV4MessageRepository.php:36]: Obviously this is a significant departure from what was previously happening here, but given that the previous constructor defined the properties as `private` it seems unlikely that anyone could extend this class in a useful way, so I doubt there is a compatibility issue.

File: src/DoctrineMessageRepository/LegacyTableSchemaTest.php [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    public function testItHoldsTableAndColumnNames(): void
+    {
+        self::assertSame('legacy_messages', $this->tableSchema->tableName());
+        self::assertSame('event_id', $this->tableSchema->eventIdColumn());
+        self::assertSame('aggregate_root_id', $this->tableSchema->aggregateRootIdColumn());
+        self::assertSame('aggregate_root_version', $this->tableSchema->versionColumn());
+        self::assertSame('payload', $this->tableSchema->payloadColumn());
+        self::assertCount(2, $this->tableSchema->additionalColumns());
+        self::assertSame(Header::TIME_OF_RECORDING, $this->tableSchema->additionalColumns()['time_of_recording']);
+        self::assertSame(Header::EVENT_TYPE, $this->tableSchema->additionalColumns()['event_type']);
```
Comment (shadowhand) [inline src/DoctrineMessageRepository/LegacyTableSchemaTest.php:30]: https://github.com/EventSaucePHP/DoctrineMessageRepository/blob/ae84fa2545230a097c7faae2670060b84e687419/tests/setup-postgres-schema.php#L11-L16

File: src/DoctrineMessageRepository/UuidSerializer.php [hunk 3]
```diff
@@ -0,0 +1,22 @@
+<?php
+
+declare(strict_types=1);
+
+namespace EventSauce\MessageRepository\DoctrineMessageRepository;
+
+use Ramsey\Uuid\Uuid;
+use Ramsey\Uuid\UuidInterface;
+
+class UuidSerializer
```
Comment (shadowhand) [inline src/DoctrineMessageRepository/UuidSerializer.php:10]: It felt really weird to have `interface UuidSerializer` for such a simple class. I am happy to modify if desired.

File: src/DoctrineMessageRepository/DoctrineMessageRepository.php [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    ) {}
+
+    public function persist(Message ...$messages): void
+    {
+        if (count($messages) === 0) {
+            return;
+        }
+
+        try {
+            $this->connection->transactional(fn () => $this->persistMessages($messages));
```
Comment (shadowhand) [inline src/DoctrineMessageRepository/DoctrineMessageRepository.php:42]: The previous implementation use string interpolation to generate a single `INSERT` query with many values. While that implementation may have been (slightly) more performant, I feel that running multiple `INSERT` statements in a transaction is just as safe and easier to understand.

File: src/DoctrineMessageRepository/DoctrineMessageRepository.php [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+        foreach ($this->tableSchema->additionalColumns() as $column => $header) {
+            $data[$column] = $payload['headers'][$header];
+        }
+
+        $data[$this->tableSchema->eventIdColumn()] = $payload['headers'][Header::EVENT_ID];
+        $data[$this->tableSchema->aggregateRootIdColumn()] = $payload['headers'][Header::AGGREGATE_ROOT_ID];
+        $data[$this->tableSchema->versionColumn()] = $payload['headers'][Header::AGGREGATE_ROOT_VERSION];
+        $data[$this->tableSchema->payloadColumn()] = json_encode($payload, $this->jsonEncodeOptions);
+
+        $this->connection->insert($this->tableSchema->tableName(), $data);
```
Comment (shadowhand) [inline src/DoctrineMessageRepository/DoctrineMessageRepository.php:120]: Using `Connection::insert()` eliminates the need to use manually create a query with string interpolation.

File: src/DoctrineMessageRepository/DoctrineMessageRepository.php [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+        $payload['headers'][Header::EVENT_ID] ??= Uuid::uuid4()->toString();
+        $payload['headers'][Header::EVENT_ID] = $this->uuidSerializer->serializeString($payload['headers'][Header::EVENT_ID]);
+        $payload['headers'][Header::AGGREGATE_ROOT_ID] = $this->uuidSerializer->serializeString($payload['headers'][Header::AGGREGATE_ROOT_ID]);
+
+        $data = [];
+
+        // Additional columns are added BEFORE known columns, so that additional
+        // columns cannot overwrite (eg) aggregate root ID column.
+        foreach ($this->tableSchema->additionalColumns() as $column => $header) {
+            $data[$column] = $payload['headers'][$header];
```
Comment (shadowhand) [inline src/DoctrineMessageRepository/DoctrineMessageRepository.php:112]: Is this comment clear?

**Comment by frankdejonge** [comment]
Hi @shadowhand,

The idea for your PR is clear. There are some things that I would like to have done differently. I'm perfectly willing to rewrite it as such. The uuid-serializer should indeed be an interface as you mentioned, and instead of having the Schema injected I'd prefer to have two top-level implementations. I can merge as it and rewrite or if you are up for it you can do it too, all the same to me :)

For future PR's (I hope there will be, I appreciate the contributions) let's try to align on some high-level design before diving into the implementation. It would also help if we tackle one issue per PR.

**Comment by shadowhand** [comment]
I'm happy to make the necessary changes.

A couple of questions:

1. What is the benefit of not having the schema injected? Given that the difference between old/new schema can be encapsulated entirely within the different TableSchema implementations?
2. What does "prefer to have two top-level implementations" mean? Do you mean have `DoctrineLegacySchemaMessageRepository` and `DoctrineNewSchemaMessageRepository`?

File: src/DoctrineMessageRepository/DoctrineMessageRepository.php (hunk 4 above)
Comment (frankdejonge) [reply src/DoctrineMessageRepository/DoctrineMessageRepository.php:42]: Was there a reason why you rewrote this with the query builder? I prefer to have one statement. At work we use these implementations as a big highway for events, the latency for the DB is multiplied by the number of events, which makes it unpredictable and thus harder to account for capacity planning.

File: src/DoctrineMessageRepository/DoctrineMessageRepository.php (hunk 6 above)
Comment (frankdejonge) [reply src/DoctrineMessageRepository/DoctrineMessageRepository.php:112]: It is.

File: src/DoctrineMessageRepository/DoctrineMessageRepository.php (hunk 4 above)
Comment (shadowhand) [reply src/DoctrineMessageRepository/DoctrineMessageRepository.php:42]: As my comment above says, I felt that a single transaction with multiple `INSERT` statements would be nearly as performant and **much** easier to understand from a code standpoint. If you feel that a single query represents a significant enough performance benefit, I am happy to return the previous implementation.

Comment (frankdejonge) [reply src/DoctrineMessageRepository/DoctrineMessageRepository.php:42]: I'd prefer that. The O(n) latency is not great for apps at the scale where Mollie operates.

Comment (shadowhand) [reply src/DoctrineMessageRepository/DoctrineMessageRepository.php:42]: Do you feel that it is important to use `:named` parameters for the query, as opposed to simple `?` placeholders? The former results in quite a few string concatenation gymnastics, which was part of the reason I switched to using `Connection::insert`.

**Comment by frankdejonge** [comment]
The only really problematic thing for the schema is the table name being included, which conceptually I can think of reasons why it would be and reasons why it would not be. If the schema is all about table names, the current implementation would remain fully BC and your case you be supported by injecting an optional TableSchema. This would remove the need for the legacy wrapping and call forwarding. If we leave it out, the table schema would be sort of static blueprints for column layout. The top-level constructor would be the same, with addition of the (optional) serializer and (again optional) table schema.

Right now, in order to maintain BC we need to keep supplying a wrapper while I think the reasons to do so are not super compelling. So the problem is not really the table schema vs two plain implementation, but more about that design quirkiness that now arrises.

File: src/DoctrineMessageRepository/DoctrineMessageRepository.php (hunk 4 above)
Comment (frankdejonge) [reply src/DoctrineMessageRepository/DoctrineMessageRepository.php:42]: I do not have a particularly strong opinion about that. There might be a param binding type difference tho, right?

**Comment by shadowhand** [comment]
> If the schema is all about table names

It isn't *only* about that, but also about allowing users to choose to extract some header values into columns for search optimization. For instance, if someone was writing the "current user ID" to the headers using a message decorator, they could have `user_id` column added to their schema to allow identifying all events that a particular user generated.

Having `TableSchema` means that there is far more flexibility, both now and in the future, for users to choose their own message storage schema (table and columns), without having to copy/paste to create their own `MessageRepository` implementation to change these details.

I think there is incredible value in this, because it would even allow making the `aggregate_root_id` column naming consistent with the underlying messages, such as `order_messages(order_id)` instead of `order_messages(aggregate_root_id)`.

Does that make sense?

**Comment by frankdejonge** [comment]
Sorry, I means column names, edited my previous comment.

**Comment by frankdejonge** [comment]
Just to make my comment clearer. If we remove the table name, only keep the column names, then we can maintain BC _and_ have the qualities you describe.

**Comment by shadowhand** [comment]
So your suggestion is to change the signature of `DoctrineUuidV4MessageRepository::__construct` to ... ?

```php
public function __construct(
    private Connection $connection,
    private string $tableName,
    private MessageSerializer $serializer,
    private TableSchema $tableSchema = null,
    private UuidEncoder $uuidEncoder = null,
) {
    $this->tableSchema ??= new DefaultTableSchema();
    $this->uuidEncoder ??= new UuidBinaryEncoder();
}
```

**Comment by frankdejonge** [comment]
Yes 👍

**Comment by shadowhand** [comment]
That still creates an (IMO) awkward constructor where some values are still strings (`$tableName`) while the rest are objects. Which has always been a pet peeve of mine; I prefer that objects either have objects for injection or non-objects for constructor parameters.

The cost of converting `DoctrineUuidV4MessageRepository` to a proxy is very low, so I don't really understand the desire for adding new constructor parameters to preserve BC. The upgrade path to `DoctrineMessageRepository` is quite direct...

**Comment by frankdejonge** [comment]
The awkwardness you describe for the DI usage is still there, only now it's pushed down one level.

**Comment by frankdejonge** [comment]
There is actually a DI benefit to leaving the table name out. In Symfony projects I can now set the default column layout / table schema to either Legacy or Default and then only need to inject the tableName string.

**Comment by shadowhand** [comment]
Okay, you've convinced me. I'll wrap up the changes and push it.

**Comment by frankdejonge** [comment]
💪 great success. Thank you for entertaining all this feedback from my end and thanks for the time and energy you're putting in.

File: src/DoctrineMessageRepository/DoctrineUuidV4MessageRepository.php [hunk 7]
```diff
@@ -3,115 +3,161 @@
...
-            $payloadColumn = 'payload_' . $index;
-            $values[] = "(:{$eventIdColumn}, :{$aggregateRootIdColumn}, :{$versionColumn}, :{$payloadColumn})";
-            $parameters[$versionColumn] = $payload['headers'][Header::AGGREGATE_ROOT_VERSION] ?? 0;
-            $payload['headers'][Header::EVENT_ID] = $payload['headers'][Header::EVENT_ID] ?? Uuid::uuid4()->toString();
-            $parameters[$eventIdColumn] = $this->uuidToBinary($payload['headers'][Header::EVENT_ID]);
-            $parameters[$payloadColumn] = json_encode($payload, $this->jsonEncodeOptions);
-            $parameters[$aggregateRootIdColumn] = $this->uuidToBinary($payload['headers'][Header::AGGREGATE_ROOT_ID] ?? '');
+            $payload['headers'][Header::EVENT_ID] ??= Uuid::uuid4()->toString();
+
+            $messageParameters = [
```
Comment (shadowhand) [inline src/DoctrineMessageRepository/DoctrineUuidV4MessageRepository.php:67]: We have to use a separate array here, rather than updating `$insertParameters`, because there is an unknown number of columns to do `TableSchema::additionalColumns()` return having a variable length.

File: src/DoctrineMessageRepository/DoctrineUuidV4MessageRepository.php [hunk 8]
```diff
@@ -3,115 +3,161 @@
...
+
+            foreach ($this->tableSchema->additionalColumns() as $column => $header) {
+                $messageParameters[$this->indexParameter(name: $column, index: $index)] = $payload['headers'][$header];
+            }
+
+            // Creates a values line like: (:event_id_1, :aggregate_root_id_1, ...)
+            $insertValues[] = implode(', ', $this->nameParameters(array_keys($messageParameters)));
+
+            // Flatten the message parameters into the query parameters
+            $insertParameters = array_merge($insertParameters, $messageParameters);
```
Comment (shadowhand) [inline src/DoctrineMessageRepository/DoctrineUuidV4MessageRepository.php:82]: I wish there was a better way to do this...

File: src/DoctrineV2MessageRepository/DoctrineUuidV4MessageRepositoryTest.php [hunk 9]
```diff
@@ -22,7 +22,6 @@ class DoctrineUuidV4MessageRepositoryTest extends MessageRepositoryTestCase
     protected function setUp(): void
     {
         if ( ! interface_exists(ResultStatement::class)) {
-            die("WHAT");
```
Comment (shadowhand) [inline src/DoctrineV2MessageRepository/DoctrineUuidV4MessageRepositoryTest.php:25]: WHAT 😆

Comment (frankdejonge) [reply src/DoctrineV2MessageRepository/DoctrineUuidV4MessageRepositoryTest.php:25]: lol, probably a left-over from my "challenges" with the multi-version setup.

File: src/DoctrineMessageRepository/DoctrineMessageRepository.php (hunk 4 above)
Comment (shadowhand) [reply src/DoctrineMessageRepository/DoctrineMessageRepository.php]: I implemented named parameters but I don't think there is a strong reason to do it, and it definitely makes generating the query more complicated. I think I'll make a separate commit that uses un-named parameters to see if it cleans things up.

File: src/DoctrineMessageRepository/DoctrineUuidV4MessageRepository.php (hunk 8 above)
Comment (frankdejonge) [reply src/DoctrineMessageRepository/DoctrineUuidV4MessageRepository.php:82]: Shouldn't this be an `array_push($insertParameters, ...$messageParameters)`? We want to append, not substitute, rigth?

Comment (shadowhand) [reply src/DoctrineMessageRepository/DoctrineUuidV4MessageRepository.php:82]: No, because the parameters are named.

File: src/DoctrineMessageRepository/DoctrineUuidV4MessageRepository.php [hunk 10]
```diff
@@ -3,115 +3,158 @@
...
+    private TableSchema $tableSchema;
+    private UuidEncoder $uuidEncoder;
 
     public function __construct(
         private Connection $connection,
         private string $tableName,
         private MessageSerializer $serializer,
-    ) {}
+        ?TableSchema $tableSchema = null,
+        ?UuidEncoder $uuidEncoder = null,
```
[the remark this answers is not in the document]
Comment (frankdejonge) [reply src/DoctrineMessageRepository/DoctrineUuidV4MessageRepository.php:40]: It'd leave constructor promotion for these for now and declare them and then use $this->tableSchema = $tableSchema ?? new DefaultTableSchema();

[the remark this answers is not in the document]
Comment (frankdejonge) [reply src/DoctrineMessageRepository/DoctrineUuidV4MessageRepository.php:40]: Ah, that's what you did. Perfect.

File: src/DoctrineMessageRepository/DoctrineUuidV4MessageRepository.php (hunk 8 above)
Comment (frankdejonge) [reply src/DoctrineMessageRepository/DoctrineUuidV4MessageRepository.php:82]: Ah right, I scanned too quickly.

**Comment by shadowhand** [comment]
After trying it, the unnamed parameter query is really ugly and hard to follow. I think what is here is better for debugging.

**Comment by frankdejonge** [comment]
Cool, let's keep it as is then 👍  almost done with the review. Will release tomorrow, almost heading to dinner.

**Comment by shadowhand** [comment]
Glad this worked out! Appreciate your willingness to expand the functionality here, @frankdejonge.

**Comment by frankdejonge** [comment]
All good! Thanks for putting in the work to get this merged. I'll be extracting some functionality, the illuminate and other doctrine package will need some shared functionality, so I'll promote the uuid encoding and the table schema to helper packages, they should be quite stable. Then I'll release them. Heading to bed now but hope I can finish this up tomorrow so you're unblocked then as well :)
